File: TheaterSuggestions/Java/ScreeningSeatingArrangementSuggestions/src/main/java/org/octo/seatingplacessuggestions/seatingplacesuggestions/SeatingPlace.java
package org.octo.seatingplacessuggestions.seatingplacesuggestions;

public record SeatingPlace(String rowName, int number, PricingCategory pricingCategory, SeatingPlaceAvailability seatingPlaceAvailability) {

    public boolean isAvailable() {
        return seatingPlaceAvailability == SeatingPlaceAvailability.AVAILABLE;
    }

    public boolean matchCategory(PricingCategory pricingCategory) {
        if (pricingCategory == PricingCategory.IGNORED) return true;
        return this.pricingCategory == pricingCategory;
    }

    public SeatingPlace allocate() {
        if (seatingPlaceAvailability == SeatingPlaceAvailability.AVAILABLE)
            return new SeatingPlace(rowName, number, pricingCategory, SeatingPlaceAvailability.ALLOCATED);
        return this;
    }

    @Override
    public String toString() {
        return rowName + number;
    }

    boolean isSameLocation(SeatingPlace seatingPlace) {
        return seatingPlace.rowName().equals(rowName()) && seatingPlace.number() == number();
    }
}

File: TheaterSuggestions/CSharp/ScreeningSeatingArrangementSuggestions/SeatingPlace.cs
namespace SeatsSuggestions;

public class SeatingPlace(
    string rowName,
    int number,
    PricingCategory pricingCategory,
    SeatingPlaceAvailability seatingPlaceAvailability) : Value.ValueType<SeatingPlace> 
{
    public string RowName { get; } = rowName;
    public int Number { get; } = number;
    public PricingCategory PricingCategory { get; } = pricingCategory;
    private SeatingPlaceAvailability SeatingPlaceAvailability { get; set; } = seatingPlaceAvailability;

    public bool IsAvailable()
    {
        return SeatingPlaceAvailability == SeatingPlaceAvailability.Available;
    }

    public override string ToString()
    {
        return $"{RowName}{Number}";
    }

    public bool MatchCategory(PricingCategory pricingCategory)
    {
        if (pricingCategory == PricingCategory.Ignored) return true;
        
        return PricingCategory == pricingCategory;
    }

    public SeatingPlace Allocate()
    {
        if (SeatingPlaceAvailability == SeatingPlaceAvailability.Available)
        {
            return new SeatingPlace(RowName, Number, PricingCategory, SeatingPlaceAvailability.Allocated);
        }
        return this;
    }

    protected override IEnumerable<object> GetAllAttributesToBeUsedForEquality()
    {
        return [RowName, Number, PricingCategory, SeatingPlaceAvailability];
    }

    public bool IsSameLocation(SeatingPlace seatingPlace)
    {
        return this.Number == seatingPlace.Number && this.RowName == seatingPlace.RowName;
    }
}